pkg/clustermanager/eksa_mover.go

```diff
@@ -0,0 +1,146 @@
+package clustermanager
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/clients/kubernetes"
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/retrier"
+	"github.com/aws/eks-anywhere/pkg/types"
+)
+
+// MoverOpt allows to customize a Mover on construction.
+type MoverOpt func(*Mover)
+
+// Mover applies the cluster spec to the management cluster and waits
+// until the changes are fully reconciled.
+type Mover struct {
+	log                logr.Logger
+	clientFactory      ClientFactory
+	moveClusterTimeout time.Duration
+	retryBackOff       time.Duration
+}
+
+// NewMover builds an Mover.
+func NewMover(log logr.Logger, clientFactory ClientFactory, opts ...MoverOpt) *Mover {
+	m := &Mover{
+		log:                log,
+		clientFactory:      clientFactory,
+		moveClusterTimeout: applyClusterSpecTimeout,
+		retryBackOff:       retryBackOff,
+	}
+
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	return m
+}
+
+// WithMoverNoTimeouts disables the timeout for all the waits and retries in management upgrader.
+func WithMoverNoTimeouts() MoverOpt {
+	return func(a *Mover) {
+		maxTime := time.Duration(math.MaxInt64)
+		a.moveClusterTimeout = maxTime
+	}
+}
+
+// WithMoverApplyClusterTimeout allows to configure how long the mover retries
+// to apply the objects in case of failure.
+// Generally only used in tests.
+func WithMoverApplyClusterTimeout(timeout time.Duration) MoverOpt {
+	return func(m *Mover) {
+		m.moveClusterTimeout = timeout
+	}
+}
+
+// WithMoverRetryBackOff allows to configure how long the mover waits between requests
+// to update the cluster spec objects and check the status of the Cluster.
+// Generally only used in tests.
+func WithMoverRetryBackOff(backOff time.Duration) MoverOpt {
+	return func(m *Mover) {
+		m.retryBackOff = backOff
+	}
+}
+
+// Move applies the cluster's namespace and spec without checking for reconcile conditions.
+func (m *Mover) Move(ctx context.Context, spec *cluster.Spec, fromCluster, toCluster *types.Cluster) error {
```

Comment:
what would you think about passing two clients instead of the cluster type?

I know this is an existing pattern but hear me out:
The only reason we pass these `type.Cluster` (afaik) is to use them to drive the configuration of a k8s client. This means that on every piece of code where we receive this and need a client, we need to inject a client factory and build the client in runntime. This leads to the same code being written again and again and having to test it again and again. I'm not concerned about the runtime performance implications but about the development overhead this adds.

Wouldn't this code (and all other similar service objects) be simpler if we just passed clients instead? And it would be simpler to test.

Opinions?

Reply:
Yeah I agree, we should do this.